File: src/main/java/com/practice/ComparatorPractice.java
package com.practice;

import java.util.HashMap;
import java.util.Map;

public class ComparatorPractice {

  public static void main(String[] args) {
    ComparatorPractice m = new ComparatorPractice();
    m.solution();
  }

  private void solution() {
    String s = "edaeaccc";
    char[] charArr = s.toCharArray();

    Map<Character, Integer> map = new HashMap<>();
    for (char c : charArr) {
      map.put(c, map.getOrDefault(c, 0)+1);
    }
    map.entrySet().forEach(System.out::println);

    var list = map.entrySet().stream()
        .sorted((o1, o2) -> {
          if (o1.getValue().equals(o2.getValue())) {
            return o1.getKey().compareTo(o2.getKey());
          } else {
            return o2.getValue().compareTo(o1.getValue());
          }
        })
        .map(Map.Entry::getKey);
    list.forEach(System.out::println);
  }
}

File: src/main/kotlin/com/practice/ComparatorPractice.kt
package com.practice

import java.util.function.Consumer
import java.util.function.Function

fun main() {
    val s = "eddddaeaccc"
    val charArr = s.toCharArray()

    val map: MutableMap<Char, Int> = HashMap()
    for (c in charArr) {
        map[c] = map.getOrDefault(c, 0) + 1
    }
    map.entries.forEach(Consumer<Map.Entry<Char, Int>> { x: Map.Entry<Char, Int>? ->
        println(
            x
        )
    })

    val list = map.entries.stream()
        .sorted { (key, value): Map.Entry<Char, Int>, (key1, value1): Map.Entry<Char, Int> ->
            if (value == value1) {
                return@sorted key.compareTo(key1)
            } else {
                return@sorted value1.compareTo(value)
            }
        }
        .map({ (key, _) -> key })
    list.forEach { x: Char? -> println(x) }
}
